### PR TITLE
Trigger a new image build when we detect that the Containerfile has changed.

### PR DESCRIPTION
**Pull request by SkrrtBacharach**
Fixes https://github.com/containers/ansible-podman-collections/issues/646

### CODE REVIEW

File: plugins/modules/podman_image.py [hunk 1]
```diff
@@ -541,15 +542,100 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
...
+        """
+
+        args_containerfile_hash = ""
+
+        context_has_containerfile = self.path and self._find_containerfile_from_context() != ""
+
+        should_hash_args_containerfile = (
+            context_has_containerfile or
+            self.build.get('file') != None or
+            self.build.get('container_file') != None
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:608]: Please fix linter issues:
```
ERROR: plugins/modules/podman_image.py:607:12: singleton-comparison: Comparison 'self.build.get('file') != None' should be 'self.build.get('file') is not None'
ERROR: plugins/modules/podman_image.py:608:12: singleton-comparison: Comparison 'self.build.get('container_file') != None' should be 'self.build.get('container_file') is not None'
```

**Comment by sshnaidm** [comment]
@SkrrtBacharach thanks for contribution! Let me review it thoroughly. Can you meanwhile please sign your commit with `git commit -s`? It's required by containers org.

**Comment by SkrrtBacharach** [comment]
> @SkrrtBacharach thanks for contribution! Let me review it thoroughly. Can you meanwhile please sign your commit with `git commit -s`? It's required by containers org.

@sshnaidm Do I have to sign with my real name/email? Would love to sign under an alias if that's possible

**Comment by sshnaidm** [comment]
> > @SkrrtBacharach thanks for contribution! Let me review it thoroughly. Can you meanwhile please sign your commit with `git commit -s`? It's required by containers org.
> 
> @sshnaidm Do I have to sign with my real name/email? Would love to sign under an alias if that's possible

No, you don't have of course, it can be any. If you can't, it's fine, I can force the job to pass :)

File: plugins/modules/podman_image.py [hunk 2]
```diff
@@ -541,15 +542,100 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
...
+            context_has_containerfile or
+            self.build.get('file') is not None or
+            self.build.get('container_file') is not None
+        )
+
+        if should_hash_args_containerfile:
+            args_containerfile_hash = self._hash_containerfile_contents(
+                self._get_containerfile_contents()
+            )
+        return args_containerfile_hash
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:615]: please fix linters:
```
ERROR: Found 1 pep8 issue(s) which need to be resolved:
ERROR: plugins/modules/podman_image.py:618:5: E303: too many blank lines (2)
```

Comment (SkrrtBacharach) [reply plugins/modules/podman_image.py:615]: Fixed in https://github.com/containers/ansible-podman-collections/pull/811/commits/a5f951303f26dca65374c96403f867382db13164

**Comment by SkrrtBacharach** [comment]
> > > @SkrrtBacharach thanks for contribution! Let me review it thoroughly. Can you meanwhile please sign your commit with `git commit -s`? It's required by containers org.
> > 
> > 
> > @sshnaidm Do I have to sign with my real name/email? Would love to sign under an alias if that's possible
> 
> No, you don't have of course, it can be any. If you can't, it's fine, I can force the job to pass :)

Done

**Comment by SkrrtBacharach** [comment]
@sshnaidm please let me know if you have questions/comments/concerns with my PR. Thanks for your help with this.

File: plugins/modules/podman_image.py [hunk 3]
```diff
@@ -491,15 +492,99 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
...
+        """
+        If we can find a Containerfile in any of the module args
+        or inside the build context, hash its contents.
+
+        If we don't have this, return an empty string.
+        """
+
+        args_containerfile_hash = ""
+
+        context_has_containerfile = self.path and self._find_containerfile_from_context() != ""
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:553]:
```suggestion
        context_has_containerfile = self.path and self._find_containerfile_from_context()
```

File: plugins/modules/podman_image.py [hunk 4]
```diff
@@ -491,15 +492,99 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
...
+
     def present(self):
         image = self.find_image()
 
+        existing_image_containerfile_hash = ""
+        args_containerfile_hash = self._get_args_containerfile_hash()
+
         if image:
             digest_before = image[0].get('Digest', image[0].get('digest'))
+            labels = image[0].get('Labels') or {}
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:575]:
```suggestion
            labels = image[0].get('Labels', {})
```

File: plugins/modules/podman_image.py [hunk 5]
```diff
@@ -698,6 +783,9 @@ def build_image(self):
                 f.write(container_file_txt)
             args.extend(['--file', container_file_path])
 
+        if containerfile_hash != "":
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:786]:
```suggestion
        if containerfile_hash:
```

File: plugins/modules/podman_image.py [hunk 6]
```diff
@@ -491,15 +492,99 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
 
         return layer_ids[-1]
 
+    def _find_containerfile_from_context(self):
+        """
+        Find a Containerfile/Dockerfile path inside a podman build context.
+        Return an empty string if none exist.
+        """
+
+        containerfile_path = ""
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:501]:
```suggestion
        containerfile_path = None
```

File: plugins/modules/podman_image.py [hunk 7]
```diff
@@ -491,15 +492,99 @@ def _get_id_from_output(self, lines, startswith=None, contains=None, split_on='
...
+
+        See if either `file` or `container_file` in build args are populated,
+        fetch their contents if so. If not, return the contents of the Containerfile
+        or Dockerfile from inside the build context.
+        """
+
+        build_file_arg = self.build.get('file') if self.build else None
+        containerfile_contents = self.build.get('container_file') if self.build else None
+
+        container_filename = ""
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:521]:
```suggestion
        container_filename = None
```

**Review by sshnaidm — COMMENTED** [review]
Thanks @SkrrtBacharach for the contribution and sorry for a such long delay. My nits are about Python style only.

File: plugins/modules/podman_image.py (hunk 7 above)
Comment (SkrrtBacharach) [reply plugins/modules/podman_image.py]: Fixed in https://github.com/containers/ansible-podman-collections/pull/811/commits/ec02a05e52ce73c30c485bd8b726d5aaab0cbe49

File: plugins/modules/podman_image.py (hunk 5 above)
Comment (SkrrtBacharach) [reply plugins/modules/podman_image.py]: Fixed in https://github.com/containers/ansible-podman-collections/pull/811/commits/ec02a05e52ce73c30c485bd8b726d5aaab0cbe49

File: plugins/modules/podman_image.py (hunk 6 above)
Comment (SkrrtBacharach) [reply plugins/modules/podman_image.py]: Fixed in https://github.com/containers/ansible-podman-collections/pull/811/commits/ec02a05e52ce73c30c485bd8b726d5aaab0cbe49

**Comment by sshnaidm** [comment]
Thanks a lot!

**Comment by SkrrtBacharach** [comment]
> Thanks a lot!

There was one change I didn't make, which was giving me a test failure. It's here: https://github.com/containers/ansible-podman-collections/pull/811#discussion_r1771980835

File: plugins/modules/podman_image.py (hunk 4 above)
Comment (SkrrtBacharach) [reply plugins/modules/podman_image.py:578]: @sshnaidm I couldn't figure out why this was giving me an error.

**Pull request merged**